### DISCUSSION TITLE
Any code to compress pdf file size? For example, a pdf with image pages, 30 pages, 30 MB, how to compress to be within 10MB?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3735] [General]
Any code to compress pdf file size? For example, a pdf with image pages, 30 pages, 30 MB, how to compress to be within 10MB?

**Comment by JorjMcKie** [comment]
Chances to achieve a compression of this magnitude without loss is an illusion - very probably.
The maximum **_lossless_** compression will be achieved by using `doc.ez_save(..., garbage=4)`.
If this does not satisfy you you need to give up some content, for example replacing all images by their gray counterparts, or even remove all images.

**Comment by nissansz** [comment]
Any code to achieve Acrobat "Reduce size" function?
I found acrobat "Reduce size" function works well.

**Reply by JorjMcKie** [reply]
nothing beyond the above mentioned.

**Comment by nissansz** [comment]
what is differene for garbage=3 or 4?
doc.ez_save(..., garbage=4).

**Reply by JorjMcKie** [reply]
3 = condense equal PDF object definitions into one, but do not check for equality of streams (= roughly binary content)
4 = also compare content streams. This entails decompression where necessary and then comparing usually large memory areas.

**Comment by nissansz** [comment]
Sometimes, this command throws errors, any solution? Original pdf may contain ocr text.

    document.subset_fonts()  


  File "D:\Python3.8.8\lib\site-packages\fitz\utils.py", line 5456, in subset_fonts
    set_old_widths(font_xref, width_table, def_width)
  File "D:\Python3.8.8\lib\site-packages\fitz\utils.py", line 5212, in set_old_widths
    doc.xref_set_key(df_xref, "DW", dwidths)
  File "D:\Python3.8.8\lib\site-packages\fitz\fitz.py", line 4249, in xref_set_key
    raise ValueError("bad 'value'")
ValueError: bad 'value'

**Reply by JorjMcKie** [reply]
Upgrading PyMuPDF will automatically use the built-in subset font function - without using fontToools.
This should solve this.

**Reply by nissansz** [reply]
> Upgrading PyMuPDF will automatically use the built-in subset font function - without using fontToools. This should solve this.

how to install this updated version?

**Reply by JorjMcKie** [reply]
What?
You never used `python -m pip install -U pymupdf`?

**Reply by nissansz** [reply]
Downloading PyMuPDF-1.24.9-cp38-none-win_amd64.whl.metadata 

Is this the version updated?

**Reply by nissansz** [reply]
It works now.

**Comment by nissansz** [comment]
Any option to remove all existing font and hidden ocr text before saving?

**Reply by JorjMcKie** [reply]
Add a full-page redaction annotation and apply it without touching anything but text (see documentation).
